File: sdk/typescript/src/cli/result-envelope.ts
export type SafetyLevel = "read" | "write" | "external" | "destructive";

export interface CliArtifact {
  artifactType: string;
  schema: string;
  producerOperation: string;
  value: Record<string, unknown>;
}

export interface CliResultEnvelope<TData = unknown> {
  ok: boolean;
  operation: string;
  summary: string;
  data: TData;
  warnings: string[];
  errors: string[];
  artifacts?: CliArtifact[];
  meta: {
    durationMs: number;
    requestId: string;
    profile?: string;
    connectorId?: string;
    system?: string;
    systemVersion?: string;
    safetyLevel?: SafetyLevel;
    retryHint?: string;
  };
}

function generateRequestId(): string {
  const ts = Date.now().toString(36);
  const rand = Math.random().toString(36).slice(2, 8);
  return `req_${ts}_${rand}`;
}

export function createCliResultEnvelope<TData = unknown>(params: {
  ok: boolean;
  operation: string;
  summary: string;
  data: TData;
  warnings?: string[];
  errors?: string[];
  artifacts?: CliArtifact[];
  durationMs: number;
  requestId?: string;
  profile?: string;
  connectorId?: string;
  system?: string;
  systemVersion?: string;
  safetyLevel?: SafetyLevel;
  retryHint?: string;
}): CliResultEnvelope<TData> {
  return {
    ok: params.ok,
    operation: params.operation,
    summary: params.summary,
    data: params.data,
    warnings: params.warnings ?? [],
    errors: params.errors ?? [],
    ...(params.artifacts && params.artifacts.length > 0 ? { artifacts: params.artifacts } : {}),
    meta: {
      durationMs: Math.max(0, Math.floor(params.durationMs)),
      requestId: params.requestId ?? generateRequestId(),
      profile: params.profile,
      connectorId: params.connectorId,
      system: params.system,
      systemVersion: params.systemVersion,
      safetyLevel: params.safetyLevel,
      retryHint: params.retryHint,
    },
  };
}

export function printCliResultEnvelope(envelope: CliResultEnvelope): void {
  console.log(JSON.stringify(envelope, null, 2));
}

File: sdk/python/centris_sdk/cli/result_envelope.py
"""Shared structured result envelope for Centris CLI commands."""

from __future__ import annotations

from dataclasses import dataclass, asdict, field
import json
import time
from typing import Any, Dict, List, Optional


SafetyLevel = str


def _request_id() -> str:
    return f"req_{int(time.time() * 1000)}"


@dataclass
class CliArtifact:
    artifact_type: str
    schema: str
    producer_operation: str
    value: Dict[str, Any] = field(default_factory=dict)


@dataclass
class CliResultMeta:
    duration_ms: int
    request_id: str = field(default_factory=_request_id)
    profile: Optional[str] = None
    connector_id: Optional[str] = None
    system: Optional[str] = None
    system_version: Optional[str] = None
    safety_level: Optional[SafetyLevel] = None
    retry_hint: Optional[str] = None


@dataclass
class CliResultEnvelope:
    ok: bool
    operation: str
    summary: str
    data: Dict[str, Any] = field(default_factory=dict)
    warnings: List[str] = field(default_factory=list)
    errors: List[str] = field(default_factory=list)
    artifacts: List[CliArtifact] = field(default_factory=list)
    meta: CliResultMeta = field(default_factory=lambda: CliResultMeta(duration_ms=0))

    def to_dict(self) -> Dict[str, Any]:
        payload = asdict(self)
        if not payload["artifacts"]:
            payload.pop("artifacts", None)
        return payload


def build_result_envelope(
    *,
    ok: bool,
    operation: str,
    summary: str,
    data: Optional[Dict[str, Any]] = None,
    warnings: Optional[List[str]] = None,
    errors: Optional[List[str]] = None,
    artifacts: Optional[List[CliArtifact]] = None,
    duration_ms: int,
    profile: Optional[str] = None,
    connector_id: Optional[str] = None,
    system: Optional[str] = None,
    system_version: Optional[str] = None,
    safety_level: Optional[SafetyLevel] = None,
    retry_hint: Optional[str] = None,
) -> CliResultEnvelope:
    return CliResultEnvelope(
        ok=ok,
        operation=operation,
        summary=summary,
        data=data or {},
        warnings=warnings or [],
        errors=errors or [],
        artifacts=artifacts or [],
        meta=CliResultMeta(
            duration_ms=max(0, int(duration_ms)),
            profile=profile,
            connector_id=connector_id,
            system=system,
            system_version=system_version,
            safety_level=safety_level,
            retry_hint=retry_hint,
        ),
    )


def emit_result_envelope(envelope: CliResultEnvelope) -> None:
    print(json.dumps(envelope.to_dict(), indent=2))
